### ISSUE TITLE
Expected type 'Union[ColumnElement[bool], _HasClauseElement, SQLCoreOperations[bool], ExpressionElementRole[bool], () -> ColumnElement[bool], LambdaElement]', got 'bool' instead 

### ISSUE THREAD

**Issue by KamalBalah** [sqlalchemy/sqlalchemy#9337]
### Describe the bug

pycharm showing this hint "Expected type 'Union[ColumnElement[bool], _HasClauseElement, SQLCoreOperations[bool], ExpressionElementRole[bool], () -> ColumnElement[bool], LambdaElement]', got 'bool' instead " and the auto complete not working after update to sqlalchemy 2+

cities = db.query(City).filter(City.city_id == city_data.city_id) 

### SQLAlchemy Version in Use

2.0.4

### DBAPI (i.e. the database driver)

psycopg2

### Database Vendor and Major Version

PostgreSQL 

### Python Version

3.11

### Operating system

Windows

### To Reproduce

```python
cities = db.query(City).filter(City.city_id == city_data.city_id)
```

### Error

```
# Expected type 'Union[ColumnElement[bool], _HasClauseElement, SQLCoreOperations[bool], ExpressionElementRole[bool], () -> ColumnElement[bool], LambdaElement]', got 'bool' instead 
```

**Comment by CaselIT** [comment]
Hi,

FYI the wrong template was selected, this is a typing issue.
Have you followed the steps outlined in that template, namely removed any stub package possibly installed (both sqlalchemy-stubs and sqlalchemy2-stubs are not compatible with v2)?

do you have that error with mypy?

**Comment by KamalBalah** [comment]
> Hi,
> 
> FYI the wrong template was selected, this is a typing issue. Have you followed the steps outlined in that template, namely removed any stub package possibly installed (both sqlalchemy-stubs and sqlalchemy2-stubs are not compatible with v2)?
> 
> do you have that error with mypy?

which template please  

this are my all packages using pip freeze
pip freeze
aiofiles==22.1.0
alembic==1.9.3       
anyio==3.6.2         
async-generator==1.10
attrs==22.2.0        
bcrypt==4.0.1        
cffi==1.15.1         
click==8.1.3         
colorama==0.4.6      
cryptography==38.0.4 
decorator==5.1.1
dnspython==2.2.1
ecdsa==0.18.0
email-validator==1.3.0
fastapi==0.92.0
greenlet==2.0.1
h11==0.14.0
h2==4.1.0
hpack==4.0.0
httptools==0.5.0
hypercorn==0.14.3
hyperframe==6.0.1
idna==3.4
Mako==1.2.4
MarkupSafe==2.1.1
outcome==1.2.0
passlib==1.7.4
priority==2.0.0
psycopg2==2.9.5
pyasn1==0.4.8
pycparser==2.21
pydantic==1.10.3
python-dotenv==0.21.0
python-jose==3.3.0
python-multipart==0.0.5
PyYAML==6.0
rsa==4.9
six==1.16.0
sniffio==1.3.0
sortedcontainers==2.4.0
SQLAlchemy==2.0.3
starlette==0.25.0
toml==0.10.2
trio==0.22.0
typing_extensions==4.4.0
uvicorn==0.20.0
validators==0.20.0
watchfiles==0.18.1
websockets==10.4
wsproto==1.2.0

**Comment by CaselIT** [comment]
The template I was referring toward was the issue template.

If you cannot reproduce with mypy I think the issue here is of pycharm and should be reported there

**Comment by KamalBalah** [comment]
> 

how to change the template  in pycharm

**Comment by CaselIT** [comment]
That's not a pycharm thing, but let's forget about it, since it does not help the issue

Can you try checking if mypy has the same issue?
With the provided example there is not enough to try and reproduce the error from this side

**Comment by KamalBalah** [comment]
> That's not a pycharm thing, but let's forget about it, since it does not help the issue
> 
> Can you try checking if mypy has the same issue? With the provided example there is not enough to try and reproduce the error from this side

mypy and vs code not showing the hint and in vs code the auto complete working in good why

**Comment by CaselIT** [comment]
you will need to provide a complete runnable example. You you provided above is of no help, since I can't run it to reproduce your issue. At the very least it would include the definition of `City`

**Comment by KamalBalah** [comment]
> you will need to provide a complete runnable example. You you provided above is of no help, since I can't run it to reproduce your issue. At the very least it would include the definition of `City`

here i create new small project 

in city.py
`
from sqlalchemy import Column, String, Integer, ForeignKey
from sqlalchemy.orm import relationship

from postgress_database import Base


class City(Base):
    __tablename__ = "cities"
    city_id = Column(Integer, primary_key=True, autoincrement=True, name="Cit_Id")
    city_name_ar = Column(String(80), nullable=False, unique=True, name="Cit_Name_Ar")
    city_name_en = Column(String(80), nullable=False, unique=True, name="Cit_Name_En")
    country_id = Column(Integer, ForeignKey("countries.Cou_Id"), nullable=False, name="Cou_Id")
    countries = relationship("Country", back_populates="cities")
`

in city_schema.py

`
from typing import Optional
from pydantic import BaseModel

class BaseCitySchema(BaseModel):
    country_id: int
    city_name_ar: str
    city_name_en: str


class UpdateCitySchema(BaseCitySchema):
    city_id: int
    country_id: Optional[int]
`

in postgress_database.py

`
from sqlalchemy import create_engine

from sqlalchemy.orm import sessionmaker, Session
from sqlalchemy.ext.declarative import declarative_base

db_driver = "postgresql://"
db_username = "postgres"
db_password = "1234"
db_host = "localhost"
db_name = "tehama"

db_url = "{0}{1}:{2}@{3}/{4}".format(db_driver, db_username, db_password, db_host, db_name)

engine = create_engine(db_url)

SessionLocal = sessionmaker(autocommit=False, autoflush=False, bind=engine)

Base = declarative_base()


def get_db():
    db: Session = SessionLocal()
    try:
        yield db
    finally:
        db.close()
`

in city_route.py

`
from typing import Dict

from fastapi import APIRouter, Depends

from city import City
from postgress_database import get_db
from city_schema import UpdateCitySchema
from sqlalchemy.orm import Session

router = APIRouter(
    prefix="/city"
)


@router.put("")
def update_route(
        city_data: UpdateCitySchema,
        db: Session = Depends(get_db)):
    cities = db.query(City).filter(City.city_id == city_data.city_id)
    city = cities.first()
    cities.update({
        City.city_name_ar: city_data.city_name_ar,
        City.city_name_en: city_data.city_name_en,
    })
    db.commit()
    db.refresh(city)
`

![image](https://user-images.githubusercontent.com/56129410/220333681-162a64e4-cbdc-4bf5-9bd6-131983d3658b.png)


the auto complete not working and  showing the hint that showing in the image

**Comment by CaselIT** [comment]
That's very much expected behaviour. You need to use `mapped_column` to have typing support.

Please refer to the documentation https://docs.sqlalchemy.org/en/20/changelog/whatsnew_20.html#migrating-an-existing-mapping

**Comment by KamalBalah** [comment]
> That's very much expected behaviour. You need to use `mapped_column` to have typing support.
> 
> Please refer to the documentation https://docs.sqlalchemy.org/en/20/changelog/whatsnew_20.html#migrating-an-existing-mapping

i change postgress_database.py

`from sqlalchemy import create_engine

from sqlalchemy.orm import sessionmaker, Session, DeclarativeBase

db_driver = "postgresql://"
db_username = "postgres"
db_password = "1234"
db_host = "localhost"
db_name = "tehama"

db_url = "{0}{1}:{2}@{3}/{4}".format(db_driver, db_username, db_password, db_host, db_name)

engine = create_engine(db_url)

SessionLocal = sessionmaker(autocommit=False, autoflush=False, bind=engine)

Base = DeclarativeBase()


def get_db():
    db: Session = SessionLocal()
    try:
        yield db
    finally:
        db.close()
`

and city.py

`from sqlalchemy import  String, Integer, ForeignKey
from sqlalchemy.orm import  mapped_column, DeclarativeBase, Mapped


class Base(DeclarativeBase):
    pass


class City(Base):
    __tablename__ = "cities"
    city_id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True, name="Cit_Id")
    city_name_ar: Mapped[str] = mapped_column(String(80), nullable=False, unique=True, name="Cit_Name_Ar")
    city_name_en: Mapped[str] = mapped_column(String(80), nullable=False, unique=True, name="Cit_Name_En")
    country_id: Mapped[int] = mapped_column(Integer, ForeignKey("countries.Cou_Id"), nullable=False, name="Cou_Id")

`
the hint not shown now but the auto complete not working well like the image  
![image](https://user-images.githubusercontent.com/56129410/220354105-a496bc03-5d68-43a2-a14f-d62e0c1bdde5.png)

**Comment by CaselIT** [comment]
That may be an issue of pycharm. Does vscode have the same issue?

**Comment by KamalBalah** [comment]
> That may be an issue of pycharm. Does vscode have the same issue?

yes same issue in vs code from any version later than SQLAlchemy > 1.4.40

**Comment by CaselIT** [comment]
I cannot reproduce the above issue
mypy or vscode will correctly report `Type of "session.query(City).filter(City.city_id == city_id)" is "Query[City]"Pylance`

**Comment by KamalBalah** [comment]
> I cannot reproduce the above issue mypy or vscode will correctly report `Type of "session.query(City).filter(City.city_id == city_id)" is "Query[City]"Pylance`


when go to definition of filter function like the image

![image](https://user-images.githubusercontent.com/56129410/220441006-5ab2bc10-3fb0-44e2-861e-f2c67e0e57bb.png)

 in version 1.4.40 shown like this 

![image](https://user-images.githubusercontent.com/56129410/220441856-d502c83e-5d2f-4b1b-b887-00e8335c33e1.png)
![image](https://user-images.githubusercontent.com/56129410/220442026-f8b9d81e-685f-4c68-8d45-01ab664b2678.png)
![image](https://user-images.githubusercontent.com/56129410/220442589-807284e4-c9d3-43c4-b835-6733cf4ebbd2.png)

but in version 2.0.4

![image](https://user-images.githubusercontent.com/56129410/220442207-6f4bf880-063f-40d0-9ae8-cf3183171ef1.png)
![image](https://user-images.githubusercontent.com/56129410/220442472-9bd077d7-1d28-4d36-81dd-1719b1085751.png)

**Comment by CaselIT** [comment]
As mentioned that seems an issue with pycharm.
I'm not sure what can sqlalchemy do here since it has no control regarding how pycharm works.

**Comment by KamalBalah** [comment]
> As mentioned that seems an issue with pycharm. I'm not sure what can sqlalchemy do here since it has no control regarding how pycharm works.

but how the version  sqlalchemy-1.4.46 working in good way i check it now and every think ok in  sqlalchemy-1.4.46 with pycharm

**Comment by CaselIT** [comment]
version 2 has vast changes in how typing work, and requires no external typing library to be present. pycharm is probably hardcoding some things to support 1.4 that does not work for v2.

**Comment by intexcor** [comment]
до сих пор не пофиксили кстати

**Comment by ra-coder** [comment]
Have the same bug

**Comment by CaselIT** [comment]
Do you have a case that reproduces using pyright or mypy?

If the issue is in pycharm you should report it to them, since we cannot test for that

**Comment by lucasmottadev** [comment]
Whoever has this error, it may be that your entity does not extend base or DeclarativeBase. With me, it was this.

```python
from sqlalchemy.orm import DeclarativeBase


class Base(DeclarativeBase):
    pass
```

**Comment by Griklit** [comment]
![image](https://github.com/user-attachments/assets/48dfd3ea-6234-49a9-ad61-78dd43e279c4)

pycharm always seems to think that the value returned by __eq__ is bool.

**Comment by Griklit** [comment]
Pycharm fixed it at 2024.2.2 

https://youtrack.jetbrains.com/issue/PY-71748/SQLAlchemy-2.0-ORM-filter-show-wrong-type-hints-in-Pycharm
